### PR TITLE
Add google/wire and inject executor service

### PR DESCRIPTION
**Pull request by nishkrishnan**
This also moves UserConfig into a config package so that other packages
can consume it in their constructors.

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -331,30 +317,15 @@ func NewServer(userConfig UserConfig, config Config) (*Server, error) {
...
-			router,
-			logger,
-			statsScope.Scope("api"),
-		)
-	}
+	router := NewRouter(underlyingRouter, parsedURL)
 
+	projectCmdOutputHandler := handlers.NewProjectCommandOutputHandler(
+		userConfig,
+		statsScope.Scope("api"),
```
Comment (Aayyush) [inline server/server.go:324]:
```suggestion
		statsScope,
```
nit: api has already been added to the scope in the initializer method here: https://github.com/lyft/atlantis/blob/080fc925c63ecfdf4b39fbe3f6a82229df62025d/server/handlers/instrumented_project_command_output_handler.go#L30

File: server/handlers/instrumented_project_command_output_handler.go [hunk 2]
```diff
@@ -14,11 +15,20 @@ type InstrumentedProjectCommandOutputHandler struct {
 	logger            logging.SimpleLogging
 }
 
-func NewInstrumentedProjectCommandOutputHandler(projectCmdOutput chan *models.ProjectCmdOutputLine,
+func NewProjectCommandOutputHandler(
+	userConfig config.UserConfig,
```
Comment (Aayyush) [inline server/handlers/instrumented_project_command_output_handler.go:19]: Should we keep the method name same?

File: server/server.go [hunk 3]
```diff
@@ -381,20 +352,27 @@ func NewServer(userConfig UserConfig, config Config) (*Server, error) {
...
 		lockingClient = locking.NewClient(boltdb)
 	}
+
 	applyLockingClient = locking.NewApplyClient(boltdb, userConfig.DisableApply)
+
+	// moved
 	workingDirLocker := events.NewDefaultWorkingDirLocker()
 
+	// moved
 	var workingDir events.WorkingDir = &events.FileWorkspace{
```
Comment (Aayyush) [inline server/server.go:376]: Can we use the provider here instead? https://github.com/lyft/atlantis/blob/080fc925c63ecfdf4b39fbe3f6a82229df62025d/server/events/working_dir.go#L72-L76 
Even though I'm guessing we'll be removing it when we move everything to wire!

**Review by Aayyush — APPROVED** [review]
Looks great! Excited to see how it'd look like when we move all of the initialization stuff to wire :)

File: server/wire.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	scope stats.Scope,
+	projectCommandOutputHandler handlers.ProjectCommandOutputHandler,
+	locker locking.Locker,
+	boltdb *db.BoltDB,
+	workingDir events.WorkingDir,
+) (*scheduled.ExecutorService, error) {
+
+	wire.Build(
+		wire.Bind(new(events.EventParsing), new(*events.EventParser)),
+		events.NewEventParser,
```
Comment (msarvar) [inline server/wire.go:34]: Do you need this here? Can't you pass in `eventParser` variable instead?

File: server/wire.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+		scheduled.NewExecutorService,
+		scheduled.NewStaleClosedPullExecutor,
+		scheduled.NewStaleOpenPullExecutor,
+		wire.Bind(new(handlers.ResourceCleaner), new(handlers.ProjectCommandOutputHandler)),
+		wire.Bind(new(vcs.GithubPullRequestGetter), new(vcs.IGithubClient)),
+		events.NewFileWorkDirIterator,
+		wire.Bind(new(events.WorkDirIterator), new(*events.FileWorkDirIterator)),
```
Comment (msarvar) [inline server/wire.go:41]: Does the ordering matter here? I couldn't find many examples for Bind in the docs.

**Review by msarvar — APPROVED** [review]
looks great overall, left couple nits

File: server/server.go (hunk 1 above)
Comment (nishkrishnan) [reply server/server.go:324]: Good call, i must have missed it.

File: server/handlers/instrumented_project_command_output_handler.go (hunk 2 above)
Comment (nishkrishnan) [reply server/handlers/instrumented_project_command_output_handler.go:19]: The same as what?

File: server/server.go (hunk 3 above)
Comment (nishkrishnan) [reply server/server.go:376]: Yeah i think i added this when i was just experimenting.

File: server/wire.go (hunk 5 above)
Comment (nishkrishnan) [reply server/wire.go:41]: nah ordering does not matter

File: server/wire.go (hunk 4 above)
Comment (nishkrishnan) [reply server/wire.go:34]: Yeah I probably should pass in the variable here otherwise multiple copies of eventParser would be created. I'll fix.

**Pull request closed without merging**